HeterogeneousCore/SonicTriton/src/triton_utils.cc

```diff
@@ -19,15 +20,33 @@ namespace triton_utils {
 
   void throwIfError(const Error& err, std::string_view msg) {
     if (!err.IsOk())
-      throw cms::Exception("TritonServerFailure") << msg << ": " << err;
+      throw cms::Exception("TritonFailure") << msg << (err.Message().empty() ? "" : ": " + err.Message());
   }
 
   bool warnIfError(const Error& err, std::string_view msg) {
     if (!err.IsOk())
-      edm::LogWarning("TritonServerWarning") << msg << ": " << err;
+      edm::LogWarning("TritonWarning") << msg << (err.Message().empty() ? "" : ": " + err.Message());
     return err.IsOk();
   }
 
+  bool warnOrThrowIfError(const Error& err, std::string_view msg, bool canThrow) {
+    if (canThrow)
+      throwIfError(err, msg);
+    return !canThrow ? warnIfError(err, msg) : err.IsOk();
+  }
+
+  void warnOrThrow(std::string_view msg, bool canThrow) {
+    warnOrThrowIfError(Error("client-side problem"), msg, canThrow);
+  }
+
+  bool cudaCheck(cudaError_t result, std::string_view msg, bool canThrow) {
```

Comment:
@makortel I meant to ask for feedback here: this duplicates part of https://github.com/cms-sw/cmssw/blob/master/HeterogeneousCore/CUDAUtilities/interface/cudaCheck.h, because I needed a version that didn't always throw (sometimes this could be called after `acquire()` returns, at which point throwing is not recommended). If it's desirable to have such a version in the main `CUDAUtilities` package, I can try to integrate it there.

Reply:
The question on "throwing destructors" is a good one. With the direct CUDA we currently have a mixed bag of throwing and non-throwing destructors (with "non-throwing" achieved by not checking the return value under the assumption that if the failure would be important a subsequent CUDA API call would fail).

The C++ Core Guidelines actually suggest to terminate the program if the destructor "has no other option than to throw" ([C.36](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-dtor-fail)). For us the downside of `terminate()` or `abort()` is that (IIUC) comp ops classify failing job by the job return value, and all jobs failing with `abort()` go to the same category and effectively need to be manually inspected.

I don't really have a good answer here. E.g. if `cudaFree()` fails, is it worth even an printout? (I can imagine cases where it would be) Also strictly speaking this `cudaCheck()` can still throw.

Reply:
It's actually not just the destructor where throwing should be avoided. `TritonOutputData::prepare()` is called in `TritonClient::evaluate()`; in `PseudoAsync` mode, `evaluate()` is called in the lightweight thread, where we don't want to throw. Also, and more generally, SONIC has a retry feature, so `evaluate()` should not throw, but rather report a failed status, in case the call should be retried.

Reply:
I'm not sure if there are valid reason for not throwing.

Even if there are, the behaviour should depend on the situation and on the error: some (most) CUDA errors are not recoverable without destroying all CUDA contexts and recreating them (which I don't think it's even possible with the runtime API, only with the driver API), so it doesn't really make sense to report a failure and retry - CUDA will simply give the same error again.

> E.g. if `cudaFree()` fails, is it worth even an printout?

What is the alternative ? Silently calling `abort()` ?

Reply:
I was under the impression that code executing after `acquire()` returns should avoid throwing (hence `edm::WaitingTask*Holder::doneWaiting()` taking an exception pointer to rethrow when back in a "normal" task).

If there's agreement that CUDA errors are unrecoverable (and, in the SONIC context, should not be retried) and it's fine to throw at any point when encountering a CUDA error, then I'll just use the existing `cudaCheck()` macro and simplify the PR somewhat.

@Dr15Jones do you have any feedback?

Reply:
> I was under the impression that code executing after `acquire()` returns should avoid throwing (hence `edm::WaitingTask*Holder::doneWaiting()` taking an exception pointer to rethrow when back in a "normal" task).

The code "launched" by `acquire()` should indeed not throw an exception, but report errors via `edm::WaitingTask*Holder::doneWaiting()`. Nevertheless it is fine to use exceptions there as long as all code possibly throwing is wrapped in a try-catch such that the catch block propagates the exception to the `doneWaiting()`.

I think we can agree that CUDA errors are largely unrecoverable (either the runtime/driver/device is messed up, or it is a user error).

Reply:
> I'm not sure if there are valid reason for not throwing.

In principle errors at the process shutdown after the framework job report has been written could be ignored as everything there is to do has been done, and aborting the process at that point would be mostly only annoying. In practice the job report appears to get finalized pretty late in the process shutdown, so the number of valid such cases is close to zero.



> > E.g. if `cudaFree()` fails, is it worth even an printout?
> 
> What is the alternative ? Silently calling `abort()` ?

In this case I specifically meant the use in a destructor in this PR (apologies for not being clear). For errors expected to be reported by the next CUDA call, the infrastructural errors in `cudaFree()` in a destructor could be ignored if the next CUDA call is done "within the same event". On the other hand, the return value should be checked for `cudaErrorInvalidValue` that is specific for that exact `cudaFree()` call. At that point the error handling would become rather complicated.

Written that, I hope infrastructural CUDA errors are "exceptional enough" that we/compops can live with jobs crashing in those cases, and user CUDA errors come up quickly enough in our testing that those get ironed out before any kind of production. Then checking the error codes in the destructors and letting them throw could be an acceptable compromise. Trying to reorganize the code such that the errors could be checked earlier than the destructor would still be better (this is where I still have some homework to do), but not always feasible.

Reply:
In that case, how about this:
1. `triton_utils::cudaCheck()` returns an exception pointer if `canThrow` is false
2. if `evaluate()` gets an exception pointer at any point, it passes it to `finish()`, which then will not retry (assuming a pre-existing exception pointer denotes an unrecoverable error) but will just give the exception pointer to `WaitingTask*Holder`
3. separately, add a `close()` accessor to `TritonMemResource` objects which will set a `closed_` status member; `close()` will only be called in the destructor if the object is not already `closed_`
4. `updateMem()` will call `close()` before calling `reset()`, so the destructor can only throw when called at the very end of the process, when we don't care.

In general, it would be nice to have a way to mock/generate CUDA errors for testing... at least, I personally don't know how to do this in a general way.

Reply:
@makortel do you have any feedback on this?

Reply:
The overall logic sounds reasonable, but I'm wondering why in 1 and 2 return the exception pointer instead of throwing and catching the exception? Naively the latter could simplify the code.

Reply:
I prefer to avoid try/catch, because it makes it harder to debug thrown exceptions in gdb.

Reply:
mhm, how so ?
I actually found it easier to catch and rethrow exactly to be able to use catch points in GDB

Reply:
>I actually found it easier to catch and rethrow exactly to be able to use catch points in GDB

That has been my experience as well.

Reply:
It's true that this facilitates `catch rethrow` for debugging rethrown exceptions specifically. However, in other circumstances, I more often want to use a simple `catch throw`. Code with a lot of try/catch blocks makes this very irritating, because gdb will stop at every unrelated try/catch block. I think my above proposal falls into the latter case - it wouldn't be an actual rethrow, because `evaluate()` should never throw.

Reply:
I would not add any `rethrow` there, but something along
```cpp
void evaluate() {
  std::exception_ptr eptr;
  try {
    // code that may throw
  } catch(...) {
    eptr = std::current_exception();
  }
  if (eptr) {
    finish(false, eptr);
  } else {
    finish(true);
  }
}
```

Reply:
@makortel thanks; after some thought, I agree that this is a better pattern. Functions throw if they need to throw, and it's up to `evaluate()` to handle exceptions as necessary. I've propagated this throughout SonicTriton for consistency. (I also added a version of `cudaCheck` in `CUDAUtilities` that handles `std::string` messages for convenience.)